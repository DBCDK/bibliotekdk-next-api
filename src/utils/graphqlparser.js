import { orderBy, uniq } from "lodash";

/**
 * Parse selections tree recursively
 *
 * @param {array} selections
 * @param {array} res
 * @param {array} path
 */
function parseSelections(selections, res = [], path = []) {
  // A selection is an object of some kind
  // we are interested in selections of kind "Field"
  // that have children (object types)
  // if the field has no children it is a simple type (string, number, etc..)
  const selectionsWithChildren = selections.filter(
    selection => selection.selectionSet && selection.selectionSet.selections
  );

  // Loop through selections that are object types
  if (selectionsWithChildren.length > 0) {
    selectionsWithChildren.forEach(selection => {
      // make copy of path
      const currentPath = [...path];
      if (selection.kind === "Field") {
        // this is a field, add its name to path
        currentPath.push(selection.name.value);
      }
      // Recursively parse the children of this field
      parseSelections(selection.selectionSet.selections, res, currentPath);
    });
  } else {
    // We are done processing the path.
    // There are no children that are object types, so we can't go deeper
    res.push(path.join(":"));
  }

  return res;
}

/**
 * Create a concise human readable key for a query
 *
 * We need a way to generate a key for a given query that is readable by a human.
 * We do this by finding paths to all "object types" (not simple types).
 * In this way queries that are similar in terms of how they request the underlying services,
 * generate the same key.
 *
 * This query:
 * {
 *   work(id: "some-workid") {
 *     title
 *     manifestations {
 *       materialType
 *     }
 *     creators {
 *       name
 *     }
 *   }
 * }
 * will generate: "QUERY_work:creators_work:manifestations"
 *
 * Note that paths to the simple types; title, materialType, and name
 * are not in the string.
 *
 * Generated string is prefixed with QUERY_
 *
 * Paths are joined by "_"
 *
 * Paths are sorted before they are joined. Queries with same fields but in different order
 * will generate same string
 *
 * Duplicates of paths are removed
 *
 *
 * @param {array} selections
 */
export function selectionsToKey(selections) {
  return `QUERY_${orderBy(uniq(parseSelections(selections))).join("_")}`;
}
