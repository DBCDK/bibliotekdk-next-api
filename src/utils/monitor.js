import { log } from "dbc-node-logger";
import client from "prom-client";

/**
 * Monitor performance of a function
 * It will record timings of calls to the given function
 *
 * @param {object} options
 * @param {string} options.name name of event to monitor
 * @param {string} options.help help text
 * @param {function} func the function to monitor
 *
 * @returns {function} wrapped function
 */
export default function monitor({ name, help }, func) {
  // Create timings histogram
  const hist = new client.Histogram({
    name,
    help,
    buckets: [0.1, 0.5, 1, 2, 5]
  });

  // Return the wrapped function
  return async function(...args) {
    // Start timer
    const end = hist.startTimer();

    try {
      return await func(...args);
    } finally {
      // End the timer
      end();
    }
  };
}

// Object that stores Prometheus Histograms
const histograms = {};

/**
 * Create a histogram for some name
 * If no histogram of the name exists, it will be created
 * If too many histograms are created, error is thrown
 * @param {string} name
 */
export function createHistogram(name) {
  let hist = histograms[name];
  if (!hist) {
    if (Object.keys(histograms).length > 99) {
      throw new Error(`Too many observations created`);
    }
    hist = new client.Histogram({
      name,
      help: "Histogram for durations",
      buckets: [0.1, 0.5, 1, 2, 5]
    });
    histograms[name] = hist;
  }
  return hist;
}

/**
 * Observe a duration for a histogram of some name
 * If no histogram of the name exists, it will be created
 * @param {string} name
 * @param {number} duration
 */
export function observeDuration(name, duration) {
  try {
    const hist = createHistogram(name);
    hist.observe(duration);
  } catch (e) {
    log.error(`Could not observe duration for name ${name}: ${e.message}`);
  }
}

const counters = {};

/**
 * Increment count by name
 * If no counter of the name exists, it will be created
 * @param {string} name
 */
export function count(name) {
  let counter = counters[name];
  if (!counter) {
    counter = new client.Counter({
      name,
      help: "A counter"
    });
    counters[name] = counter;
  }
  counter.inc();
}

/**
 * Route handler for the metrics endpoint
 *
 */
export async function metrics(req, res) {
  res.send(client.register.metrics());
}
