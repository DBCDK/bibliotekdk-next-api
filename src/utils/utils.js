import { get, uniq } from "lodash";

/**
 * Gets array at a path in some object
 * If value at path does not exists we return empty array
 * if value at path is not an array, we wrap in array
 *
 * @param {object} obj The object to look in
 * @param {string} path The path to look at
 *
 * @returns {array}
 */
export function getArray(obj, path) {
  const res = get(obj, path);
  if (res) {
    if (Array.isArray(res)) {
      return res;
    }
    return [res];
  }
  return [];
}

const regex = /\d{4}/g;
export function matchYear(str) {
  return str.match(regex);
}

/**
 * Generates the work page description
 * @param {object} work The work
 * @returns {string}
 */
export function getPageDescription({ title, creators, materialTypes }) {
  const creator = creators[0] && creators[0].name;
  const allowedTypes = ["lydbog", "ebog", "bog"];
  const types = uniq(
    materialTypes
      .map(entry => {
        for (let i = 0; i < allowedTypes.length; i++) {
          if (entry.materialType.toLowerCase().includes(allowedTypes[i])) {
            return allowedTypes[i];
          }
        }
      })
      .filter(type => !!type)
  );

  let typesString = "";
  types.forEach((type, idx) => {
    if (idx > 0) {
      if (idx === types.length - 1) {
        // last element
        typesString += " eller ";
      } else {
        // middle element
        typesString += ", ";
      }
    } else {
      // first element
      typesString = " som ";
    }
    typesString += type;
  });

  return `Lån ${title}${
    creator ? ` af ${creator}` : ""
  }${typesString}. Bestil, reserver, lån fra alle danmarks biblioteker. Afhent på dit lokale bibliotek eller find online.`;
}
