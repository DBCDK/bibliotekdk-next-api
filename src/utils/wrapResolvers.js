const SYMBOL_PROCESSED = Symbol("processed");

/**
 * Wrap every resolver within a schema with some function
 * It is useful for timing resolvers, logging etc.
 *
 * Inspired by https://github.com/kiwicom/graphql-resolve-wrapper
 *
 * @param {*} schema
 * @param {func} wrapper
 */
export function wrapResolvers(schema, wrapper) {
  const types = schema.getTypeMap();
  // Traverse types in the schema
  Object.values(types).forEach((type) => {
    if (
      type[SYMBOL_PROCESSED] ||
      !type.getFields ||
      isSystemType(type.toString())
    ) {
      return;
    }

    // Traverse every field of the type
    Object.values(type.getFields()).forEach((field) => {
      // The original resolve function
      const resolveFn = field.resolve;

      if (field[SYMBOL_PROCESSED] || !resolveFn) {
        return;
      }

      field[SYMBOL_PROCESSED] = true;

      // Wrap the resolve function
      field.resolve = wrapper(resolveFn, field);
    });
  });
}

function isSystemType(fieldName) {
  // __TypeKind, __InputValue, ...
  return /^__/.test(fieldName);
}
