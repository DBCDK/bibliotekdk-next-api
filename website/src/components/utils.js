var months = [
  "Jan",
  "Feb",
  "Mar",
  "Apr",
  "May",
  "Jun",
  "Jul",
  "Aug",
  "Sep",
  "Oct",
  "Nov",
  "Dec",
];

export function dateTimeConverter(timestamp) {
  var a = new Date(timestamp);
  var year = a.getFullYear();
  var month = months[a.getMonth()];
  var date = a.getDate();
  var hour = a.getHours() < 10 ? "0" + a.getHours() : a.getHours();
  var min = a.getMinutes() < 10 ? "0" + a.getMinutes() : a.getMinutes();
  var sec = a.getSeconds() < 10 ? "0" + a.getSeconds() : a.getSeconds();
  var time =
    date + ". " + month + ". " + year + " " + hour + ":" + min + ":" + sec;
  return time;
}

export function isToken(token) {
  // alpha numeric and more than 32 characters
  return !!(token && token.match(/^(?=.*[a-zA-Z])(?=.*[0-9]).{40}/));
}
