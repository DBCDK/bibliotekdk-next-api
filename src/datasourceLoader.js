import DataLoader from "dataloader";
import { log } from "dbc-node-logger";
import { withRedis } from "./datasources/redis.datasource";
import monitor from "./utils/monitor";
import { getFilesRecursive } from "./utils/utils";

// Find all datasources in src/datasources
export const datasources = getFilesRecursive("./src/datasources")
  .map((file) => {
    if (!file.path.endsWith(".datasource.js")) {
      return;
    }
    const { load, options, batchLoader, status } = require(file.path);
    if (!load) {
      return;
    }

    // Extract datasource name from filename
    const name = file.file.replace(".datasource.js", "");

    // Monitor the load function from the datasource
    const monitoredLoad = monitor(
      {
        name: `REQUEST_${name}`,
        help: `${name} requests`,
      },
      load
    );

    // if datasource exports a createBatchLoader we use that,
    // otherwise we use default batch loader
    let monitoredBatchLoader = batchLoader
      ? (keys) => batchLoader(keys, monitoredLoad)
      : async (keys) => {
          return await Promise.all(keys.map((key) => monitoredLoad(key)));
        };

    // Check if Redis is configured for this datasource
    if (options && options.redis && options.redis.prefix && options.redis.ttl) {
      monitoredBatchLoader = withRedis(monitoredBatchLoader, {
        prefix: options.redis.prefix,
        ttl: options.redis.ttl,
      });
    }

    const statusChecker = status && (async () => await status(monitoredLoad));

    return {
      batchLoader: monitoredBatchLoader,
      name,
      options,
      statusChecker,
    };
  })
  .filter((func) => !!func);

log.info(`found ${datasources.length} datasources`, {
  datasources: datasources.map((datasource) => ({
    name: datasource.name,
    options: datasource.options,
  })),
});

/**
 * Will instantiate dataloaders from datasources
 */
export default function createDataLoaders() {
  const result = {};
  datasources.forEach((datasource) => {
    result[datasource.name] = new DataLoader(datasource.batchLoader, {
      // If key is an object, we stringify
      // to make it useful as a cache key
      cacheKeyFn: (key) =>
        typeof key === "object" ? JSON.stringify(key) : key,
    });
  });
  return result;
}
