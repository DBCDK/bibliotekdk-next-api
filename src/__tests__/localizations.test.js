import { createMockedDataLoaders } from "../datasourceLoader";
import { performTestQuery } from "../utils/utils";

test("localizations - get for a number of pids", async () => {
  const result = await performTestQuery({
    query: `
          query  {
            work(id: "work-of:870970-basis:26521556") {
              materialTypes {          
                localizations{
                  count
                  agencies{
                  agencyId
                    holdingItems{
                      localizationPid
                      localIdentifier
                      codes
                    }
                  }
                }
              }
            }
          }
        `,
    variables: {},
    context: { datasources: createMockedDataLoaders() },
  });

  console.log(result, "RESULT");

  expect(result).toMatchSnapshot();
});
