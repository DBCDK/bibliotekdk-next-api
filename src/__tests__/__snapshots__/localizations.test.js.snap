// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`localizations - get for a number of pids 1`] = `
Object {
  "data": Object {
    "work": Object {
      "materialTypes": Array [
        Object {
          "localizations": null,
        },
        Object {
          "localizations": null,
        },
        Object {
          "localizations": null,
        },
        Object {
          "localizations": null,
        },
        Object {
          "localizations": null,
        },
        Object {
          "localizations": null,
        },
      ],
    },
  },
  "errors": Array [
    [GraphQLError: Cannot read property 'admindata' of undefined],
    [GraphQLError: Cannot read property 'admindata' of undefined],
    [GraphQLError: Cannot read property 'admindata' of undefined],
    [GraphQLError: Cannot read property 'admindata' of undefined],
    [GraphQLError: Cannot read property 'admindata' of undefined],
    [GraphQLError: Cannot read property 'admindata' of undefined],
  ],
}
`;
